dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/StreamJsonParser.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json.Nodes;
+
+namespace Microsoft.SemanticKernel.Connectors.GoogleVertexAI.Core;
+
+/// <summary>
+/// Internal class for parsing a stream of jsons into separate json objects.
+/// </summary>
+internal sealed class StreamJsonParser
+{
+    /// <summary>
+    /// Parses a Stream containing JSON data and yields the individual JSON objects.
+    /// </summary>
+    /// <param name="stream">The Stream containing the JSON data.</param>
+    /// <param name="validateJson">Set to true to enable JSON validation. Default is false.</param>
+    /// <returns>An enumerable collection of string representing the individual JSON objects.</returns>
+    public IEnumerable<string> Parse(Stream stream, bool validateJson = false)
+    {
+        using var reader = new StreamReader(stream, Encoding.UTF8);
+        while (ExtractNextJsonObject(reader, validateJson) is { } json)
+        {
+            yield return json;
+        }
+    }
+
+    private static string? ExtractNextJsonObject(TextReader reader, bool validateJson)
+    {
+        JsonParserState state = new();
+        while ((state.CharacterInt = reader.Read()) != -1)
```

Comment:
It is quite universal parser. Gemini returns chunk as fully json objects. Each of chunks sometimes contain one or even few lines of text.

Reply:
> It is quite universal parser.

That's my concern. It's implemented doing synchronous I/O and processing one character at a time. That means it'll be blocking even for asynchronous consumers, and rather than doing vectorized searching (e.g. what IndexOfAny does), it's instead having to examine each character individually.

Reply:
I will analyse this and let you what I can do with that